### PR TITLE
[subscriptions] Implement new template file for subscription PDP frequency selector

### PR DESCRIPTION
**Pull request by smahbod2014**
# Description
This PR contains all the necessary changes to add a subscription frequency selector to the product details page.

[Template file task](https://app.asana.com/0/1202967843415720/1203093658712998)
[Magento admin setting task](https://app.asana.com/0/1202967843415720/1203093658713000)

Screen Shot 2022-11-17 at 1 25 15 PM

This is a core functionality for our Subscription team's MVP in January for First-part Physical Subscription offerings.

**This PR contains 3 sections:**
## 1. Adding a new Magento config setting of type `YesNo`
Whether to display the product details page frequency selector. This is the flag that our beta merchants will manually enable to see the rest of the changes, and consequently see the frequency selector on their subscribable product detail pages.

Changes: 
- Define and added the config in system.xml
- `getPDPFrequencySelectorFlag()` in Config.php that checks for that flag
- Create `isPDPFrequencySelectorEnabled()` that checks if the above config returns `true`, and `isBoltProductPageForFrequencySelector()` that checks for the prior and if the Shopper is currently on the product page.

## 2. Loading a new .phtml template into the product details page
We want the plugin to load a new file into the PDP page and place it in a similar position to the existing product page checkout button. We will call the template `subscription_freq_selector_product_page.phtml`.

We also want to enable loading the BoltConnect script when `isBoltProductPageForFrequencySelector()` is true

Changes:
- Add a `block in catalog_product_view.xml` in the `addtocart` section, for the template to be placed in that area of the product details page.
- update `$isLoadBoltJs` load condition to include `isBoltProductPageForFrequencySelector()`

## 3. Create the `subscription_freq_selector_product_page.phtml`template
Create a new template that encapsulates all the rendering/script loading for this frequency selector.

Notes:
- It checks for `isBoltProductPageForFrequencySelector()` before doing anything else.
- It has a new empty selector `<div class="bolt-freq-selector"></div>` that we will inject the freq selector into
- It loads the bolt script in a similar manner as the product checkout button does
- it gets product details similar to the product checkout button
- there is one callback `onSubscriptionChanged` which is called when the Shopper selects a frequency.
- It loads the BoltConnect script.

Fixes: (link ticket)

#changelog [subscriptions] Implement new template file for subscription PDP frequency selector

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 

- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my ticket link and provided a changelog message.

### CODE REVIEW

File: view/frontend/templates/subscription_freq_selector_product_page.phtml [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        var getQty = function () {
+            var quantity = Number($('#qty').val());
+            return quantity > 0 ? quantity : 1;
+        }
+
+        var getGroupedProductChildQty = function () {
+            return Number($('input[name="super_group['+productId+']').val());
+        }
+
+        var getItemsData = function () {
```
Comment (Nitsorn) [inline view/frontend/templates/subscription_freq_selector_product_page.phtml:82]: All of these came from https://github.com//BoltApp/bolt-magento2/blob/master/view/frontend/templates/button_product_page.phtml#L211 - all used to get the items on this page. We need the items on the BoltConnect side to identify if the product(s) are subscribable.

File: view/frontend/templates/subscription_freq_selector_product_page.phtml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+        var cart;
+        var setupProductPage = function () {
+            cart = {
+                items: getItemsData()
+            };
+
+            // if connect.js is not loaded postpone until it is
+            whenDefined(window, 'BoltCheckout', function(){
+                BoltCheckout.configureProductFrequencySelector(cart, null, callbacks);
```
Comment (Nitsorn) [inline view/frontend/templates/subscription_freq_selector_product_page.phtml:177]: We will be sending a a PR to create `configureProductFrequencySelector()` on Source/storm side very soon!

**Review by garmashdev — APPROVED** [review]

File: etc/adminhtml/system.xml [hunk 3]
```diff
@@ -93,6 +93,12 @@
                 <group id="additional" translate="label" type="text" sortOrder="50" showInDefault="1" showInWebsite="1" showInStore="1">
                     <label>Additional options</label>
                     <attribute type="expanded">1</attribute>
+                    <field id="pdp_frequency_selector" translate="label" type="select" sortOrder="80" showInDefault="1" showInWebsite="1" showInStore="0">
```
Comment (garmashdev) [inline etc/adminhtml/system.xml:96]: please add default value to "config.xml"

**Review by garmashdev — APPROVED** [review]

File: view/frontend/templates/subscription_freq_selector_product_page.phtml (hunk 2 above)
Comment (vitaliyreznikov) [reply view/frontend/templates/subscription_freq_selector_product_page.phtml:177]: @Nitsorn Is it ok to merge the changes in storm first and then this one?
We are not to release during code freeze so we have some time.

Comment (Nitsorn) [reply view/frontend/templates/subscription_freq_selector_product_page.phtml:177]: yes definitely, we are in the process of getting those in. We certainly have time, but want you to take a look, even if we dont merge, to know that we are on the right path.

Comment (vitaliyreznikov) [reply view/frontend/templates/subscription_freq_selector_product_page.phtml:177]: Sure. Our team will check during next few days

File: etc/adminhtml/system.xml (hunk 3 above)
Comment (smahbod2014) [reply etc/adminhtml/system.xml:96]: Added

File: view/frontend/templates/subscription_freq_selector_product_page.phtml [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            return quantity > 0 ? quantity : 1;
+        }
+
+        var getGroupedProductChildQty = function () {
+            return Number($('input[name="super_group['+productId+']').val());
+        }
+
+        var getItemsData = function () {
+            var options = {};
+                var compositeAttributes = ['super_attribute', 'options', 'bundle_option', 'bundle_option_qty'];
```
Comment (vitaliyreznikov) [inline view/frontend/templates/subscription_freq_selector_product_page.phtml:84]: Is there any way to reduce code duplication between PPC and subscription?

File: view/frontend/templates/subscription_freq_selector_product_page.phtml (hunk 2 above)
Comment (Nitsorn) [reply view/frontend/templates/subscription_freq_selector_product_page.phtml:177]: please and thank you!

File: view/frontend/templates/subscription_freq_selector_product_page.phtml (hunk 4 above)
Comment (smahbod2014) [reply view/frontend/templates/subscription_freq_selector_product_page.phtml:84]: Is it ok if we merge this and followup with a cleanup PR for reducing code duplication?

File: view/frontend/templates/subscription_freq_selector_product_page.phtml [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+             * This callback is triggered by the frequency selector's onChange
+             * User has selected a different frequency
+             * 
+             * We update the localStorage value that is accessible when shopper checks out
+             * LocalStorage value is a dictionary with
+             * - key as merchant product id ("reference" here)
+             * - value is the selected frequency (value, unit eg: 2, 'weeks')
+             */
+            onSubscriptionChanged: function(reference, frequency) {
+                console.log("subscription frequency changed! ", reference, frequency);
```
Comment (vitaliyreznikov) [inline view/frontend/templates/subscription_freq_selector_product_page.phtml:54]: Do we need this line?

**Review by vitaliyreznikov — APPROVED** [review]

File: view/frontend/templates/subscription_freq_selector_product_page.phtml (hunk 5 above)
Comment (smahbod2014) [reply view/frontend/templates/subscription_freq_selector_product_page.phtml]: Removed

**Pull request merged**